src/Evolution/Systems/Cce/LinearSolve.hpp

```diff
@@ -0,0 +1,322 @@
+  using integrand_tags =
+      tmpl::list<Tags::PoleOfIntegrand<Tags::BondiH>,
+                 Tags::RegularIntegrand<Tags::BondiH>,
+                 Tags::LinearFactor<Tags::BondiH>,
+                 Tags::LinearFactorForConjugate<Tags::BondiH>>;
+  using integration_independent_tags = tmpl::list<Tags::OneMinusY>;
+
+  using return_tags = tmpl::list<Tags::BondiH>;
+  using argument_tags =
+      tmpl::append<integrand_tags, boundary_tags, integration_independent_tags,
+                   tmpl::list<Spectral::Swsh::Tags::LMax,
+                              Spectral::Swsh::Tags::NumberOfRadialPoints>>;
+  static void apply(
+      const gsl::not_null<Scalar<SpinWeighted<ComplexDataVector, 2>>*>
+          integral_result,
+      const Scalar<SpinWeighted<ComplexDataVector, 2>>& pole_of_integrand,
+      const Scalar<SpinWeighted<ComplexDataVector, 2>>& regular_integrand,
+      const Scalar<SpinWeighted<ComplexDataVector, 0>>& linear_factor,
+      const Scalar<SpinWeighted<ComplexDataVector, 4>>&
+          linear_factor_of_conjugate,
+      const Scalar<SpinWeighted<ComplexDataVector, 2>>& boundary,
+      const Scalar<SpinWeighted<ComplexDataVector, 0>>& one_minus_y,
+      const size_t l_max, const size_t number_of_radial_points) noexcept {
+    const size_t number_of_angular_points =
+        Spectral::Swsh::number_of_swsh_collocation_points(l_max);
+
+    Matrix operator_matrix(2 * number_of_radial_points,
+                           2 * number_of_radial_points);
+
+    ComplexDataVector integrand =
+        get(pole_of_integrand).data() +
+        get(one_minus_y).data() * get(regular_integrand).data();
+
+    DataVector transpose_buffer{2 * get(pole_of_integrand).size()};
+    DataVector linear_solve_buffer{2 * get(pole_of_integrand).size()};
+
+    // transpose such that each radial slice is split up into the order:
+    // (real radial slice 00) (imag radial slice 00) (real radial slice 01) ...
+    detail::transpose_to_reals_then_imags_radial_stripes(
+        make_not_null(&linear_solve_buffer), integrand, number_of_radial_points,
+        number_of_angular_points);
+
+    raw_transpose(make_not_null(transpose_buffer.data()),
+                  linear_solve_buffer.data(), number_of_radial_points,
+                  2 * number_of_angular_points);
+
+    const auto mesh = Mesh<1>{{{number_of_radial_points}},
+                              Spectral::Basis::Legendre,
+                              Spectral::Quadrature::GaussLobatto};
+    const auto& derivative_matrix =
+        Spectral::differentiation_matrix(mesh.slice_through(0));
+    for (size_t offset = 0; offset < number_of_angular_points; ++offset) {
+      // on repeated evaluations, the matrix gets permuted by the dgesv routine.
+      // We'll ignore its pivots and just overwrite the whole thing on each
+      // pass. There are probably optimizations that can be made which make use
+      // of the pivots.
+      for (size_t chunk = 0; chunk < 2; ++chunk) {
```

Comment:
chunk  -> matrix_block and add comments describing what blocks are being touched by which code